doc/api/url.md

```diff
@@ -2,139 +2,244 @@
+* Otherwise, if `urlObject.protocol` is not `undefined` and is not a string, an
+  [`Error`][] is thrown.
+* For all string values of `urlObject.protocol` that *do not end* with an ASCII
+  colon (`:`) character, the literal string `:` will be appended to `result`.
+* If either the `urlObject.slashes` property is true, `urlObject.protocol`
+  begins with one of `http`, `https`, `ftp`, `gopher`, or `file`, or
+  `urlObject.protocol` is `undefined`, the literal string `//` will be appended
+  to `result`.
+* If the value of the `urlObject.auth` property is truthy, and either
+  `urlObject.host` or `urlObject.hostname` are not `undefined`, the value of
+  `urlObject.auth` will be coerced into a string and appended to `result`
+   followed by the literal string `@`.
+* If the `urlObject.host` property is `undefined` then:
+  * If the `urlObject.hostname` is a string, it is appended to `result`.
+  * Otherwise, if `urlObject.hostname` is not `undefined` and is not a string,
+    an [`Error`][] is thrown.
+  * If the `urlObject.port` property value is truthy, and `urlObject.hostname`
+    is not `undefined`:
+    * The literal string `:` is appended to `result`, and
+    * The value of `urlObject.port` is coerced to a string and appended to
+      `result`.
+* Otherwise, if the `urlObject.host` property value is truthy, the value of
+  `urlObject.host` is coerced to a string and appended to `result`.
+* If the `urlObject.pathname` property is a string that is not an empty string:
+  * If the `urlObject.pathname` *does not start* with an ASCII forward slash
+    (`/`), then the literal string '/' is appended to `result`.
+  * The value of `urlObject.pathname` is appended to `result`.
+* Otherwise, if `urlObject.pathname` is not `undefined` and is not a string, an
+  [`Error`][] is thrown.
+* If the `urlObject.search` property is `undefined` and if the `urlObject.query`
+  property is an `Object`, the literal string `?` is appended to `result`
+  followed by the output of calling the `querystring` module's `stringify()`
+  function passing the value of `urlObject.query`.
+* Otherwise, if `urlObject.search` is a string:
+  * If the value of `urlObject.search` *does not start* with the ASCII question
+    mark (`?`) character, the literal string `?` is appended to `result`.
+  * The value of `urlObject.search` is appended to `result`.
+* Otherwise, if `urlObject.search` is not `undefined` and is not a string, an
+  [`Error`][] is thrown.
+* If the `urlObject.hash` property is a string:
+  * If the value of `urlObject.hash` *does not start* with the ASCII hash (`#`)
+    character, the literal string `#` is appended to `result`.
+  * The value of `urlObject.hash` is appended to `result`.
+* Otherwise, if the `urlObject.hash` property is not `undefined` and is not a
+  string, an [`Error`][] is thrown.
+* `result` is returned.
+
 
 ## url.parse(urlStr[, parseQueryString][, slashesDenoteHost])
```

Comment:
This really needs to be clarified since it's not possible to specify `slashesDenoteHost` without specifying `parseQueryString` first: `urlStr[, parseQueryString[, slashesDenoteHost]]`